pkg/operator/ceph/csi/csidriver_object.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func (b1csi beta1CsiDriver) createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name, fsGroupPolicy string) error {
+	attach := true
+	mountInfo := false
+	// Create CSIDriver object
+	csiDriver := &betav1k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: betav1k8scsi.CSIDriverSpec{
+			AttachRequired: &attach,
+			PodInfoOnMount: &mountInfo,
+		},
+	}
+	if fsGroupPolicy != "" {
+		policy := betav1k8scsi.FSGroupPolicy(fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
+	csidrivers := clientset.StorageV1beta1().CSIDrivers()
+	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
+	if err == nil {
+		// As FSGroupPolicy field is immutable, should be set only during create time.
+		// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+		if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
+			return b1csi.reCreateCSIDriverInfo(ctx)
+		}
+
+		// For csidriver we need to provide the resourceVersion when updating the object.
+		// From the docs (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata)
+		// > "This value MUST be treated as opaque by clients and passed unmodified back to the server"
+		csiDriver.ObjectMeta.ResourceVersion = driver.ObjectMeta.ResourceVersion
+		_, err = csidrivers.Update(ctx, csiDriver, metav1.UpdateOptions{})
+		if err == nil {
+			logger.Infof("CSIDriver object updated for driver %q", name)
+		}
+		return err
+	}
+
+	if apierrors.IsNotFound(err) {
+		_, err = csidrivers.Create(ctx, csiDriver, metav1.CreateOptions{})
+		if err == nil {
+			logger.Infof("CSIDriver object created for driver %q", name)
+		}
+	}
+
+	return err
+}
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func (v1csi v1CsiDriver) createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name, fsGroupPolicy string) error {
+	attach := true
+	mountInfo := false
+	// Create CSIDriver object
+	csiDriver := &v1k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1k8scsi.CSIDriverSpec{
+			AttachRequired: &attach,
+			PodInfoOnMount: &mountInfo,
+		},
+	}
+	if fsGroupPolicy != "" {
+		policy := v1k8scsi.FSGroupPolicy(fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
+	csidrivers := clientset.StorageV1().CSIDrivers()
+	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
+	if err == nil {
+		// As FSGroupPolicy field is immutable, should be set only during create time.
+		// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+		if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
```

Comment:
It might be nice to factor out some of this common logic into helper methods to avoid duplication. But since there is so much duplication anyway it may not be worth it. wdyt?

Reply:
I was thinking the same but not sure that helps because we only keep adding the new parameters to the `v1` spec. and not do any changes to `betav1` apart from any bug fixes. as we remove betav1 functions later IMO we should be fine for now.

Reply:
Okay :+1: